Move Noah MP init to CCPP and add new variables for Noah MP

climbfuji opened the pull request:
## Description

This PR does several things:

1. Move the Noah MP initialization of surface fields from fv3atm (`io/FV3GFS_io.F90`) to CCPP from @climbfuji.
2. Updates to Noah MP and Noah MP - radiation coupling from @helinwei-noaa. The changes in ccpp-physics require adding four additional variables in `GFS_typedefs.F90`.
3. Revert a change in `ccpp/data/GFS_typedefs.F90`, made as part of the UGWP update PR (https://github.com/NOAA-EMC/fv3atm/pull/241), that permanently disabled the old GFS v12-v15 convective gravity wave drag scheme. @SMoorthi-emc @ValeryYudin-NOAA please check, around line 4030 in the updated `GFS_typedefs.F90`.
4. Cleanup work in `GFS_typdefs.F90` for the recently merged RRTMGP updates: Add `Interstitial%gas_concentrations%concs(:)%conc` to the `Interstitial%rad_reset` call, remove RRTMGP variables that no longer exist from `GFS_typedefs.meta`, add missing active flags to RRTMGP variables in `GFS_typedefs.meta`

### Issue(s) addressed

This finally completes step 2 of https://github.com/NOAA-EMC/fv3atm/issues/214.

Fixes https://github.com/NOAA-EMC/fv3atm/issues/206.

## Testing

For regression testing, see https://github.com/ufs-community/ufs-weather-model/pull/425.

## Dependencies

https://github.com/NCAR/ccpp-physics/pull/575
https://github.com/NOAA-EMC/fv3atm/pull/249
https://github.com/ufs-community/ufs-weather-model/pull/425

climbfuji committed:
Update .gitmodules and submodule pointer for ccpp-physics for code review and testing

climbfuji committed:
Move Noah MP initialization to CCPP, add/update variables and variable metadata in ccpp/data/GFS_typedefs.{F90,meta}

climbfuji committed:
Remove gfsphysics

climbfuji committed:
ccpp/driver/GFS_restart.F90: required changes for restart reproducibility with NoahMP

climbfuji committed:
Fix cmake dependencies

HelinWei-NOAA committed:
updating noah mp

climbfuji committed:
Update submodule pointer for ccpp-physics

climbfuji committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into noahmp_init_in_ccpp

climbfuji committed:
Fix indentation in ccpp/data/GFS_typedefs.F90

climbfuji committed:
Update submodule pointer for ccpp-physics, fix formatting in ccpp/data/GFS_typedefs.F90

junwang-noaa commented on CMakeLists.txt:
```diff
@@ -109,7 +108,9 @@ target_include_directories(fv3dycore INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT
                                                $<INSTALL_INTERFACE:mod>)
 
 target_link_libraries(fv3dycore PUBLIC fms
-                                       gfsphysics
```
I am curious, we only had noahmp_tables.f90 in gfsphysics directory, does fv3dycore have dependency on that file? Can we remove this line?

climbfuji replied:
The dycore has dependencies on CCPP because of the fast physics calls (saturation adjustment) in `fv_mapz.F90`. It won't compile w/o these dependencies. The `gfsphysics` line had the CCPP dependencies included implicitly because of line 126 below in the old code.

junwang-noaa replied:
So fv3dycore has no dependency on gfsphysics itself, previous code lists gfsphysics not ccpp as dependency, maybe ccpp physics can be listed here explicitly? Also do we need ccpp? It seems the dependency is on ccppdata and ccppphys.

climbfuji replied:
The new code has dependencies ccppdata (for the data structures), ccppphys for the fast physics code, and ccpp (=ccpp-framework) for the CCPP API:
```
target_link_libraries(fv3dycore PUBLIC fms
                                       ccpp
                                       ccppphys
                                       ccppdata
                                       esmf)
```

junwang-noaa replied:
Sorry I don't get the dependency of ccpp (ccpp-framwork). I run the GFSv16 test without the line of ccpp, the code runs fine. 
 target_link_libraries(fv3dycore PUBLIC fms
 \-                                       ccpp
                                        ccppphys
                                        ccppdata
                                        esmf)

climbfuji replied:
I can try. Maybe it's yet another implicit dependency in ccppphys.

climbfuji replied:
It works without that line, too. But I am wondering if it is the right thing to do instead of spelling out all dependencies explicitly. @DusanJovic-NOAA what do you think?

DusanJovic-NOAA replied:
If `fv3dycore` does not depend directly on `ccpp` (ie. ccpp-framwork), then I would leave it out. BTW at some point we should rename it from `ccpp` to `ccppframework` to be more explicit about what it is.

climbfuji replied:
The dycore does depend explicitly on ccpp=ccppframework, because it uses `call ccpp_physics_init` etc.

DusanJovic-NOAA replied:
In that case, specify `ccpp` as explicit dependency.

climbfuji replied:
Ok, thanks.

junwang-noaa commented:
In the compile log file, the ccpp_static_api.F90 with ccpp_physics_init and
ccpp_physics_run are build in  FV3/ccpp/driver/libccppdriver.a, not in
FV3/ccpp/framework/src/libccpp.a. So it seems not from ccpp framework

On Wed, Feb 24, 2021 at 9:20 AM Dom Heinzeller <notifications@github.com>
wrote:

> *@climbfuji* commented on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/NOAA-EMC/fv3atm/pull/249#discussion_r581997034>:
>
> > @@ -109,7 +108,9 @@ target_include_directories(fv3dycore INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT
>                                                 $<INSTALL_INTERFACE:mod>)
>
>  target_link_libraries(fv3dycore PUBLIC fms
> -                                       gfsphysics
>
> The dycore does depend explicitly on ccpp=ccppframework, because it uses call
> ccpp_physics_init etc.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/fv3atm/pull/249#discussion_r581997034>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TJYA4JPL3KPDXFBO7TTAUDK5ANCNFSM4X67L6LQ>
> .
>

climbfuji commented:
But they use the ccpp type ccpp_t, which is defined in FV3/ccpp/framework/src/ccpp_types.F90 !

> On Feb 24, 2021, at 7:34 AM, Jun Wang <notifications@github.com> wrote:
> 
> 
> In the compile log file, the ccpp_static_api.F90 with ccpp_physics_init and
> ccpp_physics_run are build in FV3/ccpp/driver/libccppdriver.a, not in
> FV3/ccpp/framework/src/libccpp.a. So it seems not from ccpp framework
> 
> On Wed, Feb 24, 2021 at 9:20 AM Dom Heinzeller <notifications@github.com>
> wrote:
> 
> > *@climbfuji* commented on this pull request.
> > ------------------------------
> >
> > In CMakeLists.txt
> > <https://github.com/NOAA-EMC/fv3atm/pull/249#discussion_r581997034>:
> >
> > > @@ -109,7 +108,9 @@ target_include_directories(fv3dycore INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT
> > $<INSTALL_INTERFACE:mod>)
> >
> > target_link_libraries(fv3dycore PUBLIC fms
> > - gfsphysics
> >
> > The dycore does depend explicitly on ccpp=ccppframework, because it uses call
> > ccpp_physics_init etc.
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/NOAA-EMC/fv3atm/pull/249#discussion_r581997034>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AI7D6TJYA4JPL3KPDXFBO7TTAUDK5ANCNFSM4X67L6LQ>
> > .
> >
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/NOAA-EMC/fv3atm/pull/249#issuecomment-785117680>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB5C2RJGNGPE6CXGEY5NQQLTAUE7PANCNFSM4X67L6LQ>.
>

climbfuji commented:
```
use CCPP_data,         only: ccpp_suite,          &
                                cdata => cdata_tile, &
                                CCPP_interstitial
```
where cdata is from ccpp-framework (ccpp_t). Is this a direct or indirect dependency? Whatever, I'll remove the one line.

junwang-noaa commented:
I think that is implicit, fv3 dycore itself only needs ccpp_data and
ccpp_static_api

On Wed, Feb 24, 2021 at 9:39 AM Dom Heinzeller <notifications@github.com>
wrote:

> use CCPP_data,         only: ccpp_suite,          &
>                                 cdata => cdata_tile, &
>                                 CCPP_interstitial
>
> where cdata is from ccpp-framework (ccpp_t). Is this a direct or indirect
> dependency? Whatever, I'll remove the one line.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/fv3atm/pull/249#issuecomment-785121055>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AI7D6TNO4KF3UT7IGNHISJDTAUFQ3ANCNFSM4X67L6LQ>
> .
>

climbfuji commented:
> I think that is implicit, fv3 dycore itself only needs ccpp_data and ccpp_static_api
> […](#)

Agreed, I'll make this change and address @barlage's comments to the ccpp-physics PR when it these PRs are the next to commit.

climbfuji committed:
Remove ccpp (=ccpp-framework) from target_link_libraries of the fv3dycore

climbfuji committed:
Merge branch 'develop' of https://github.com/NOAA-EMC/fv3atm into noahmp_init_in_ccpp

climbfuji committed:
Revert a change in ccpp/data/GFS_typedefs.F90, made as part of the UGWP updates, that permanently disabled the old GFS v12-v15 convective gravity wave drag scheme

climbfuji committed:
Update submodule pointer for ccpp-physics

climbfuji committed:
Add Interstitial%gas_concentrations%concs(:)%conc to Interstitial%rad_reset call, remove RRTMGP variables from GFS_typedefs.meta that no longer exist, add missing active flags to RRTMGP variables in GFS_typedefs.meta

climbfuji commented:
@dustinswales I added the rad_reset call for the gas concentrations that was missing in the last PR, and fixed a few RRTMGP issues in GFS_typedefs.meta. Please check commit 1d9c8487e2ebc765fd44df8fada49508fd733869. Thanks!

climbfuji committed:
Update submodule pointer for ccpp-physics

SMoorthi-emc reviewed (approved):
I assume the changes in radiation_surface are applicable only to NoahMP.

climbfuji commented:
> I assume the changes in radiation_surface are applicable only to NoahMP.

Yes, they are only used when the `ialbflg` parameter is set to 2. All regression tests that do not use Noah MP pass against the existing baselines with those changes.

climbfuji committed:
Revert change to .gitmodules and update submodule pointer for fv3atm
